test/create.js

```diff
@@ -76,6 +77,36 @@ test('custom headers (extend)', async t => {
 	t.is(headers.unicorn, 'rainbow');
 });
 
+test('extend overwrites arrays', t => {
+	const statusCodes = [408];
+	const a = got.extend({retry: {statusCodes}});
+	t.deepEqual(a.defaults.options.retry.statusCodes, statusCodes);
+	t.is(a.defaults.options.retry.statusCodes, statusCodes);
+});
+
+test('extend removes object values set to undefined', t => {
+	const a = got.extend({
+		headers: {foo: 'foo', bar: 'bar', baz: 'baz'}
+	});
+	const b = a.extend({headers: {foo: undefined, bar: null}});
+	t.deepEqual(
+		b.defaults.options.headers,
+		{...got.defaults.options.headers, ...{baz: 'baz', bar: null}}
+	);
+});
+
+test('extend merges URL instances', t => {
+	const a = got.extend({baseUrl: new URL('https://example.com')});
+	const b = a.extend({baseUrl: '/foo'});
+	t.is(b.defaults.options.baseUrl.toString(), 'https://example.com/foo');
+});
+
+test('extend removes object values set to undefined (root keys)', t => {
+	t.true('headers' in got.defaults.options);
+	const a = got.extend({headers: undefined});
+	t.false('headers' in a.defaults.options);
+});
```

Comment:
:-1: It should keep the old value.

Reply:
I'm fine either way on this one.

I interpreted the previous behavior as `undefined` being a sentinel value that can be used to remove a setting, i.e. a discrete feature of got's extension behavior.

There is a test that verifies that null headers are omitted from the normalized arguments, but I've addressed that case in normalize-arguments.

@sindresorhus do you have a preference?

Reply:
> Just to clarify, based on your complete suggestion below, you think undefined values in the source should be noop and the target value should be preserved, right?

yes.

Reply:
Both arguments are valid and both are common. `Object.assign`/object-spread will overwrite with `undefined`, while many extends libs will not. I definitely find it more convenient when `{foo: true}, {foo: undefined}` end up as `{foo: true}`, but that also means there's no way to explicitly unset a value, like with headers. Maybe we do a middle ground and say that `null` overrides, while `undefined` does not. This is how default function arguments work. They only work with `undefined` not `null`. So to unset a header you set `null`.

```
{foo: true}, {foo: false}
// {foo: false}

{foo: true}, {foo: undefined}
// {foo: true}

{foo: true}, {foo: null}
// {foo: null}
```

Thoughts?

Reply:
I strongly agree :) This behaviour is more clear IMO.

Reply:
@jstewmon Could you allow edits from maintainers? I did what @sindresorhus said, I just wanted to push one commit. You'll change that (or revert) if you'd like to.

Reply:
I lightly object b/c ignoring undefined precludes the possibility of removing a setting (hypothetically, null may be a significant of invalid value). I can't think of a reason why that would be a problem today. But, I also can't think of a reason why an object would contain a key with the value set to undefined in this context b/c the key could simply be omitted to get the same behavior.

In other words, the key's absence or presence is not significant if the value is undefined, which is not something I'd expect as a user. If I set the key in the passed options, I did so with a purpose.

@szmarczak , I backed off your change b/c it overreached what we had discussed. I don't mind maintainer commits for nits or agreed upon changes, but please don't make behavioral changes when consensus has not been reached.

I think the only significant change I reverted was the deep merging of arrays. I did that because got does not make changes to those arrays, so there is no need to make a copy of them, so copying adds unnecessarily to the runtime of the method.

Reply:
We're trying to build consensus around these changes. Making a fork at this point is counterproductive.

If there's a change you want to see, we should talk about it.

Reply:
Already did.

Reply:
You 👍 my explanation for why arrays are directly assigned, then changed the behavior in a push to my branch.

You rewrote the docs I wrote without having provided any remarks.

You implemented @sindresorhus proposition on my branch without giving me a chance to respond.

I can appreciate that you are eager to land this, but we've had a productive conversation and are very close to achieving consensus on the implementation. By forging ahead, you undercut that process and then we have to have these unproductive meta conversations.

What's the rationale for, or problem solved, by deep merging arrays?

Reply:
> What's the rationale for, or problem solved, by deep merging arrays?

You don't even simply clone them, so let's start with that. With arrays there should be no exception. I'm not gonna explain simple things.

I really appreciate you want to do it on your own :)

Reply:
> I'm not gonna explain simple things.

Arguments are worthless if the other part doesn't understand you. Not everyone is in your exact mindset. We all need to be on the same page. I was also confused by your comment.

Reply:
I just realized that my proposal is how `lodash.merge` works.

> But, I also can't think of a reason why an object would contain a key with the value set to undefined in this context b/c the key could simply be omitted to get the same behavior.

It allows you to do inline conditionals:

```js
foo({
	unicorn: checkForUnicorns() && '🦄'
});
```

Without, I would have to do:

```js
const options = {};

if (shouldUseUnicorns()) {
	options.unicorn = '🦄';
}

foo(options);
```

> I backed off your change b/c it overreached what we had discussed. I don't mind maintainer commits for nits or agreed upon changes, but please don't make behavioral changes when consensus has not been reached.

@szmarczak Don't modify other people's PRs before we have agreed on a solution.

Reply:
@sindresorhus thanks for providing that example. I can appreciate the convenience that provides.

It occurred to me that, in the future, we could expose a symbol support deleting a key if the need for it arises. Similarly, we could expose a symbol to indicate that the result should be `undefined` if there's a need to explicitly set a key to `undefined`.

Using symbols for those cases would have the disadvantage of being a unique approach, but it would have the advantage of our interface being optimized for the most common scenarios.

If we're settled on ignoring `undefined`, we could rename `assignOptions` to `mergeOptions` to help clarify that its behavior is not congruent with Object.assign.

Reply:
> Arguments are worthless if the other part doesn't understand you. Not everyone is in your exact mindset. We all need to be on the same page. I was also confused by your comment.

I'll sum up: going through shortcuts won't end well, just a hint. Another hint: references, arrays, deep freeze. This says a lot.

> Don't modify other people's PRs before we have agreed on a solution.

I've just proposed a solution based on yours. Do what you want, I'm off. I'll be happy with any (current one is cool too).